File: python-legacy/callsign_snap.py
"""CallsignSnap: deterministic post-ASR callsign correction against a live
candidate list (ADS-B in-range traffic + filed flight plan).

WHY: error_analysis.py showed callsign failure is a DIGIT problem — both
Whisper and the CTC models nearly always get the airline telephony word right
and garble the flight number. Snapping the extracted callsign to the nearest
candidate (when unambiguous) cut false callsigns 14%->2% (whisper-small-us)
and 43%->2% (zipformer) in simulation. This module is the production
reference for that stage; port to Swift alongside ATCNormalize/ATCCorrector.

Two output channels, used differently downstream:
  * TEXT: the transcript with the callsign span rewritten IF a confident
    unique snap exists. Unverified callsigns stay as heard (we never delete
    what the pilot may want to see) — display-layer channel.
  * ENTITY verdict: verified_exact / snapped / unverified / no_callsign.
    Only verified/snapped callsigns may be attributed to an aircraft
    (CallsignExtractor -> ADS-B match). "unverified" = abstain — this is
    where the falseCS -> ~2% win comes from.

Candidates are canonical spoken-telephony strings ("delta 232",
"november 345 alpha bravo"). Build them from ADS-B flights via the airline
telephony map; never feed raw ICAO codes or registrations (the correction
validator's deny-list lesson, build 21).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional, Sequence, Tuple

from atc_diarize import extract_callsign, _normalize_for_match
from atc_normalize import normalize as _canon


def _lev(a: str, b: str) -> int:
    if a == b:
        return 0
    prev = list(range(len(b) + 1))
    for i, ca in enumerate(a, 1):
        cur = [i]
        for j, cb in enumerate(b, 1):
            cur.append(min(prev[j] + 1, cur[j - 1] + 1, prev[j - 1] + (ca != cb)))
        prev = cur
    return prev[-1]


def _split_cs(cs: str) -> Tuple[str, str]:
    """Canonical callsign -> (telephony word, number/letter remainder, no spaces)."""
    parts = cs.split()
    return parts[0], "".join(parts[1:])


def match_callsign(
    cs: str,
    candidates: Sequence[str],
    max_airline_ed: int = 2,
    max_num_ed: int = 1,
) -> Optional[str]:
    """Nearest unambiguous candidate for a canonical callsign, else None.

    Distance = 2*edit(telephony word) + edit(number block); a candidate
    qualifies only within (max_airline_ed, max_num_ed). Ties between two
    real aircraft mean genuine ambiguity -> abstain. Inputs may be in any
    digit format; everything is compared in atc_normalize's canonical
    per-digit space ("delta 2 3 2"), and the canonical match is returned.

    Known limitation: a misheard TELEPHONY word ("dominair" for an airline
    not in the map) never reaches this function — the extractor only anchors
    on known telephony words, so those errors surface as missed, not false.
    Gold taxonomy says that's rare (0-1 of 51); the digit-garble case this
    fixes is 7-21 of 51.
    """
    ha, hn = _split_cs(_canon(cs))
    scored = []
    for c in candidates:
        c = _canon(c)  # tolerate un-canonicalized inputs ("delta 232")
        ca, cn = _split_cs(c)
        da, dn = _lev(ha, ca), _lev(hn, cn)
        if da <= max_airline_ed and dn <= max_num_ed:
            scored.append((2 * da + dn, c))
    if not scored:
        return None
    scored.sort()
    if len(scored) > 1 and scored[0][0] == scored[1][0]:
        return None
    return scored[0][1]


@dataclass
class SnapEdit:
    """Outcome of one snap attempt. `verdict` drives the entity channel."""

    verdict: str                 # verified_exact | snapped | unverified | no_callsign
    original: Optional[str] = None   # canonical callsign as heard
    snapped: Optional[str] = None    # canonical callsign after snap (attribution-safe)
    applied: bool = False            # True iff the TEXT was rewritten


def snap_transcript(
    text: str,
    candidates: Sequence[str],
    max_airline_ed: int = 2,
    max_num_ed: int = 1,
) -> Tuple[str, SnapEdit]:
    """Return (possibly rewritten text, SnapEdit).

    The returned text is in normalized-token form (lowercase, no punctuation)
    — the same space the corrector pipeline already works in.
    """
    norm = _normalize_for_match(text)
    span = extract_callsign(norm)
    if not span:
        return text, SnapEdit(verdict="no_callsign")

    heard = _canon(span)
    match = match_callsign(heard, candidates, max_airline_ed, max_num_ed)
    if match is None:
        return text, SnapEdit(verdict="unverified", original=heard)
    if match == heard:
        return text, SnapEdit(
            verdict="verified_exact", original=heard, snapped=match)

    tokens = norm.split()
    stoks = span.split()
    for i in range(len(tokens) - len(stoks) + 1):
        if tokens[i: i + len(stoks)] == stoks:
            new_tokens = tokens[:i] + match.split() + tokens[i + len(stoks):]
            return " ".join(new_tokens), SnapEdit(
                verdict="snapped", original=heard, snapped=match, applied=True)
    # span found by the extractor but not relocatable verbatim — do no harm
    return text, SnapEdit(verdict="unverified", original=heard)

File: ios/ATCTranscribe/Core/CallsignSnap.swift
import Foundation

/// CallsignSnap — deterministic callsign correction against a live candidate list (fresh ADS-B
/// traffic in telephony form). Swift port of `python-legacy/callsign_snap.py`, validated
/// byte-identical via `snap_fixtures.json` (see `SnapParityTests`).
///
/// Measured on the US gold set: false callsign attributions 13.7% → 2.0% (whisper-small-us).
/// Two channels: the TEXT is rewritten only on a confident unique snap (unverified stays exactly
/// as heard — display honesty); the `verdict` gates aircraft ATTRIBUTION (unverified = abstain)
/// and feeds the confidence gate + LLM grounding.
enum CallsignSnap {

    struct Result: Sendable, Equatable {
        /// verified_exact | snapped | unverified | no_callsign
        let verdict: String
        var original: String?     // canonical callsign as heard
        var snapped: String?      // canonical callsign after snap (attribution-safe)
        var applied = false       // true iff the text was rewritten
    }

    /// Spoken aviation digit words + grouping cardinals (mirror of `atc_diarize._DIGIT_WORDS`).
    /// "fourty" is an ATCNormalize-only alias absent from the Python reference set — removed
    /// for byte-parity of extraction spans.
    static let digitWords: Set<String> = Set(ATCNormalize.units.keys)
        .union(ATCNormalize.teens.keys).union(ATCNormalize.tens.keys)
        .union(["hundred", "thousand"]).subtracting(["fourty"])

    /// ICAO phonetic letters (mirror of `atc_diarize._PHONETIC_WORDS`).
    static let phoneticWords: Set<String> = [
        "alpha", "bravo", "charlie", "delta", "echo", "foxtrot", "golf", "hotel",
        "india", "juliet", "juliett", "kilo", "lima", "mike", "november", "oscar",
        "papa", "quebec", "romeo", "sierra", "tango", "uniform", "victor", "whiskey",
        "xray", "x-ray", "yankee", "zulu",
    ]

    /// Lowercase telephony first-words that anchor an airline callsign ("delta", "united", …).
    /// Built from the knowledge base at call time so Python (airport_context) and Swift
    /// (airlines.json) stay behaviorally aligned; a small embedded fallback keeps the stage
    /// usable if the knowledge resource is missing.
    static func telephonyWords(_ knowledge: ATCKnowledgeBase?) -> Set<String> {
        var names: Set<String> = ["american", "delta", "united", "southwest", "jetblue",
                                  "spirit", "frontier", "alaska", "skywest", "fedex", "ups"]
        for name in knowledge?.allTelephonyNames ?? [] {
            if let first = name.lowercased().split(separator: " ").first { names.insert(String(first)) }
        }
        return names
    }

    /// Lowercase + strip to ASCII a-z/0-9 spaced tokens (mirror of `_normalize_for_match`,
    /// whose regex keeps only `[a-z0-9\s]` — non-ASCII Whisper output must tokenize the same
    /// way on both sides).
    static func normalizeForMatch(_ text: String) -> String {
        var s = ""
        for ch in text.lowercased() {
            s.append((("a"..."z").contains(ch) || ("0"..."9").contains(ch) || ch == " ") ? ch : " ")
        }
        return s.split(separator: " ").joined(separator: " ")
    }

    /// Best-effort spoken-callsign span (mirror of `atc_diarize.extract_callsign`): an airline
    /// telephony word or "november" followed by a run of digit words / numerals / phonetics.
    static func extractCallsign(_ text: String, telephony: Set<String>) -> String? {
        let tokens = normalizeForMatch(text).split(separator: " ").map(String.init)
        guard !tokens.isEmpty else { return nil }

        func isDigitish(_ t: String) -> Bool { digitWords.contains(t) || t.allSatisfy(\.isNumber) && !t.isEmpty }
        func run(from idx: Int) -> String {
            var out = [tokens[idx]]
            var j = idx + 1
            while j < tokens.count, isDigitish(tokens[j]) || phoneticWords.contains(tokens[j]) {
                out.append(tokens[j]); j += 1
            }
            return out.joined(separator: " ")
        }

        for (i, tok) in tokens.enumerated() where tok == "november" && i + 1 < tokens.count {
            if isDigitish(tokens[i + 1]) || phoneticWords.contains(tokens[i + 1]) { return run(from: i) }
        }
        for (i, tok) in tokens.enumerated() where telephony.contains(tok) && i + 1 < tokens.count {
            if isDigitish(tokens[i + 1]) { return run(from: i) }
        }
        return nil
    }

    /// Nearest unambiguous candidate for a canonical callsign, else nil (mirror of
    /// `match_callsign`): distance = 2·edit(telephony word) + edit(number block); ties between
    /// two real aircraft mean genuine ambiguity → abstain. Everything compares in
    /// `ATCNormalize` canonical per-digit space; the canonical match is returned.
    static func matchCallsign(_ cs: String, candidates: [String],
                              maxAirlineEd: Int = 2, maxNumEd: Int = 1) -> String? {
        let (ha, hn) = splitCallsign(ATCNormalize.normalize(cs))
        var scored: [(Int, String)] = []
        for raw in candidates {
            let c = ATCNormalize.normalize(raw)
            let (ca, cn) = splitCallsign(c)
            let da = levenshtein(ha, ca), dn = levenshtein(hn, cn)
            if da <= maxAirlineEd, dn <= maxNumEd { scored.append((2 * da + dn, c)) }
        }
        guard !scored.isEmpty else { return nil }
        scored.sort { $0.0 < $1.0 || ($0.0 == $1.0 && $0.1 < $1.1) }
        if scored.count > 1, scored[0].0 == scored[1].0 { return nil }
        return scored[0].1
    }

    /// Snap the transcript's callsign (mirror of `snap_transcript`). Returns the text in
    /// normalized-token space (the corrector pipeline's working space) plus the verdict.
    static func snapTranscript(_ text: String, candidates: [String],
                               telephony: Set<String>) -> (text: String, result: Result) {
        let norm = normalizeForMatch(text)
        guard let span = extractCallsign(norm, telephony: telephony) else {
            return (text, Result(verdict: "no_callsign"))
        }
        let heard = ATCNormalize.normalize(span)
        guard let match = matchCallsign(heard, candidates: candidates) else {
            return (text, Result(verdict: "unverified", original: heard))
        }
        if match == heard {
            return (text, Result(verdict: "verified_exact", original: heard, snapped: match))
        }
        // SECURITY (red-hat 2026-07-07): the candidate list comes from UNAUTHENTICATED ADS-B
        // (airplanes.live) — anyone can inject a ghost aircraft one digit off from a real one.
        // So the digit-changing rewrite path is disabled: the pilot-visible callsign DIGITS are
        // never invented from traffic. A snap may fix only the misheard AIRLINE WORD / phonetics
        // (digits identical to a live aircraft); if the digits differ, we display as heard and
        // do NOT attribute (verdict "unverified"). Digit correction is the exclusive domain of
        // guarded stages, never a single spoofable source.
        let (_, heardNum) = splitCallsign(heard)
        let (_, matchNum) = splitCallsign(match)
        guard heardNum == matchNum else {
            return (text, Result(verdict: "unverified", original: heard))
        }
        let tokens = norm.split(separator: " ").map(String.init)
        let sToks = span.split(separator: " ").map(String.init)
        if sToks.count <= tokens.count {
            for i in 0...(tokens.count - sToks.count) where Array(tokens[i..<(i + sToks.count)]) == sToks {
                let rebuilt = (tokens[0..<i] + match.split(separator: " ").map(String.init) + tokens[(i + sToks.count)...])
                    .joined(separator: " ")
                return (rebuilt, Result(verdict: "snapped", original: heard, snapped: match, applied: true))
            }
        }
        // span found by the extractor but not relocatable verbatim — do no harm
        return (text, Result(verdict: "unverified", original: heard))
    }

    /// Canonical callsign → (telephony word, number/letter remainder without spaces).
    static func splitCallsign(_ cs: String) -> (String, String) {
        let parts = cs.split(separator: " ").map(String.init)
        guard let first = parts.first else { return ("", "") }
        return (first, parts.dropFirst().joined())
    }

    static func levenshtein(_ a: String, _ b: String) -> Int {
        if a == b { return 0 }
        let aa = Array(a), bb = Array(b)
        var prev = Array(0...bb.count)
        for (i, ca) in aa.enumerated() {
            var cur = [i + 1]
            cur.reserveCapacity(bb.count + 1)
            for (j, cb) in bb.enumerated() {
                cur.append(min(prev[j + 1] + 1, cur[j] + 1, prev[j] + (ca == cb ? 0 : 1)))
            }
            prev = cur
        }
        return prev[bb.count]
    }
}
